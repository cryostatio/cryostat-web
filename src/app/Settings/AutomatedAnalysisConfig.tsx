/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { AutomatedAnalysisConfigForm } from '@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisConfigForm';
import { ServiceContext } from '@app/Shared/Services/Services';
import { AutomatedAnalysisRecordingConfig } from '@app/Shared/Services/Settings.service';
import { TargetView } from '@app/TargetView/TargetView';
import {
  DescriptionList,
  DescriptionListDescription,
  DescriptionListGroup,
  DescriptionListTerm,
  ExpandableSection,
  Stack,
  StackItem,
  Title,
} from '@patternfly/react-core';
import * as React from 'react';
import { UserSetting } from './Settings';

export const defaultAutomatedAnalysisRecordingConfig: AutomatedAnalysisRecordingConfig = {
  templates: 'template=Continuous,type=TARGET',
  maxSize: 2048,
  maxAge: 0,
};

const Component = () => {
  const context = React.useContext(ServiceContext);
  const [config, setConfig] = React.useState<AutomatedAnalysisRecordingConfig>(defaultAutomatedAnalysisRecordingConfig);
  const [expanded, setExpanded] = React.useState(false);

  React.useEffect(() => {
    const c = context.settings.automatedAnalysisRecordingConfig();
    setConfig(c);
  }, []);

  const onSave = React.useCallback(() => {
    const newConfig = context.settings.automatedAnalysisRecordingConfig();
    setConfig(newConfig);
  }, [context.settings, context.settings.automatedAnalysisRecordingConfig, setConfig]);

  return (
    <TargetView pageTitle={''} compactSelect>
      <Stack hasGutter>
        <StackItem>
          <Title headingLevel="h2" size="lg">
            Current configuration
          </Title>
        </StackItem>
        <StackItem>
          <DescriptionList columnModifier={{ lg: '3Col' }}>
            <DescriptionListGroup>
              <DescriptionListTerm>Template</DescriptionListTerm>
              <DescriptionListDescription>{config.templates}</DescriptionListDescription>
            </DescriptionListGroup>
            <DescriptionListGroup>
              <DescriptionListTerm>Max Size (B)</DescriptionListTerm>
              <DescriptionListDescription>{config.maxSize}</DescriptionListDescription>
            </DescriptionListGroup>
            <DescriptionListGroup>
              <DescriptionListTerm>Max Age (s)</DescriptionListTerm>
              <DescriptionListDescription>{config.maxAge}</DescriptionListDescription>
            </DescriptionListGroup>
          </DescriptionList>
        </StackItem>
        <ExpandableSection
          toggleText={expanded ? 'Show less' : 'Edit configuration'}
          onToggle={setExpanded}
          isExpanded={expanded}
        >
          <AutomatedAnalysisConfigForm onSave={onSave} isSettingsForm={true} />
        </ExpandableSection>
      </Stack>
    </TargetView>
  );
};

export const AutomatedAnalysisConfig: UserSetting = {
  title: 'Automated Analysis Recording Configuration',
  description:
    'Set the recording configuration for automated analysis recordings. You may want smaller or larger values for max-age and max-size depending on how recent you want events to be recorded from the analysis.',
  content: Component,
};
