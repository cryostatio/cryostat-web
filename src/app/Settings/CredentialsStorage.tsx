/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import * as React from 'react';
import { Link } from 'react-router-dom';
import { Select, SelectOption, SelectVariant, Text } from '@patternfly/react-core';
import { UserSetting } from './Settings';
import { getFromLocalStorage, saveToLocalStorage } from '@app/utils/LocalStorage';

export interface Location {
  key: string;
  description: string;
}

export class Locations {
  static readonly BROWSER_SESSION: Location = {
    key: 'Session (Browser Memory)',
    description:
      'Keep credentials in browser memory for the current session only. When you close this browser tab the credentials will be forgotten.',
  };
  static readonly BACKEND: Location = {
    key: 'Backend',
    description:
      'Keep credentials in encrypted Cryostat backend storage. These credentials will be available to other users and will be used for Automated Rules.',
  };
}

const locations = [Locations.BROWSER_SESSION, Locations.BACKEND];

const getLocation = (key: string): Location => {
  for (let l of locations) {
    if (l.key === key) {
      return l;
    }
  }
  return Locations.BACKEND;
};

const Component = () => {
  const [isExpanded, setExpanded] = React.useState(false);
  const [selection, setSelection] = React.useState(Locations.BACKEND.key);

  const handleSelect = React.useCallback(
    (_, selection) => {
      let location = getLocation(selection);
      setSelection(location.key);
      setExpanded(false);
      saveToLocalStorage('JMX_CREDENTIAL_LOCATION', selection);
    },
    [getLocation, setSelection, setExpanded, saveToLocalStorage]
  );

  React.useEffect(() => {
    handleSelect(undefined, getFromLocalStorage('JMX_CREDENTIAL_LOCATION', Locations.BACKEND.key));
  }, [handleSelect, getFromLocalStorage]);

  return (
    <>
      <Select
        variant={SelectVariant.single}
        onToggle={setExpanded}
        onSelect={handleSelect}
        isOpen={isExpanded}
        selections={selection}
      >
        {locations.map((location) => (
          <SelectOption key={location.key} value={location.key} description={location.description} />
        ))}
      </Select>
    </>
  );
};

export const CredentialsStorage: UserSetting = {
  title: 'JMX Credentials Storage',
  description: (
    <Text>
      When you attempt to connect to a target application which requires authentication, you will see a prompt for
      credentials to present to the application and complete the connection. You can choose where to persist these
      credentials. Any credentials added through the <Link to="/security">Security</Link> panel will always be stored in
      Cryostat backend encrypted storage.
    </Text>
  ),
  content: Component,
};
