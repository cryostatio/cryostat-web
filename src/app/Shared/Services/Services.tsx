/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { NotificationsInstance } from '@app/Notifications/Notifications';
import { TargetService, TargetInstance } from './Target.service';
import { TargetsService } from './Targets.service';
import { ApiService } from './Api.service';
import { NotificationChannel } from './NotificationChannel.service';
import { ReportService } from './Report.service';
import { SettingsService } from './Settings.service';
import { LoginService } from './Login.service';
import { JmxCredentials } from './JmxCredentials.service';

export interface Services {
  target: TargetService;
  targets: TargetsService;
  api: ApiService;
  jmxCredentials: JmxCredentials;
  notificationChannel: NotificationChannel;
  reports: ReportService;
  settings: SettingsService;
  login: LoginService;
}

const settings = new SettingsService();
const jmxCredentials = new JmxCredentials(() => api);
const login = new LoginService(TargetInstance, jmxCredentials, settings);
const api = new ApiService(TargetInstance, NotificationsInstance, login);
const notificationChannel = new NotificationChannel(NotificationsInstance, login);
const reports = new ReportService(login, NotificationsInstance);
const targets = new TargetsService(api, NotificationsInstance, login, notificationChannel);

const defaultServices: Services = {
  target: TargetInstance,
  targets,
  api,
  jmxCredentials,
  notificationChannel,
  reports,
  settings,
  login,
};

const ServiceContext: React.Context<Services> = React.createContext(defaultServices);

export { ServiceContext, defaultServices };
