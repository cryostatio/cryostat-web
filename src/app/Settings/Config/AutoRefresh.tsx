/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DurationPicker } from '@app/DurationPicker/DurationPicker';
import { ServiceContext } from '@app/Shared/Services/Services';
import { Checkbox } from '@patternfly/react-core';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { SettingTab, UserSetting } from '../types';

const defaultPreferences = {
  autoRefreshEnabled: true,
  autoRefreshPeriod: 30,
  autoRefreshUnits: 1_000,
};

const Component = () => {
  const { t } = useTranslation();
  const context = React.useContext(ServiceContext);
  const [state, setState] = React.useState(defaultPreferences);

  React.useLayoutEffect(() => {
    setState({
      autoRefreshEnabled: context.settings.autoRefreshEnabled(),
      autoRefreshPeriod: context.settings.autoRefreshPeriod(),
      autoRefreshUnits: context.settings.autoRefreshUnits(),
    });
  }, [setState, context.settings]);

  const handleAutoRefreshEnabledChange = React.useCallback(
    (autoRefreshEnabled) => {
      setState((state) => ({ ...state, autoRefreshEnabled }));
      context.settings.setAutoRefreshEnabled(autoRefreshEnabled);
    },
    [setState, context.settings],
  );

  const handleAutoRefreshPeriodChange = React.useCallback(
    (autoRefreshPeriod) => {
      setState((state) => ({ ...state, autoRefreshPeriod }));
      context.settings.setAutoRefreshPeriod(autoRefreshPeriod);
    },
    [setState, context.settings],
  );

  const handleAutoRefreshUnitScalarChange = React.useCallback(
    (autoRefreshUnits) => {
      setState((state) => ({ ...state, autoRefreshUnits }));
      context.settings.setAutoRefreshUnits(autoRefreshUnits);
    },
    [setState, context.settings],
  );

  return (
    <>
      <DurationPicker
        enabled={state.autoRefreshEnabled}
        period={state.autoRefreshPeriod}
        onPeriodChange={handleAutoRefreshPeriodChange}
        unitScalar={state.autoRefreshUnits}
        onUnitScalarChange={handleAutoRefreshUnitScalarChange}
      />
      <Checkbox
        id="auto-refresh-enabled"
        label={t('SETTINGS.AUTO_REFRESH.CHECKBOX_LABEL')}
        isChecked={state.autoRefreshEnabled}
        onChange={handleAutoRefreshEnabledChange}
        data-quickstart-id="settings-connectivity-tab-auto-refresh"
      />
    </>
  );
};

export const AutoRefresh: UserSetting = {
  titleKey: 'SETTINGS.AUTO_REFRESH.TITLE',
  descConstruct: 'SETTINGS.AUTO_REFRESH.DESCRIPTION',
  content: Component,
  category: SettingTab.CONNECTIVITY,
};
