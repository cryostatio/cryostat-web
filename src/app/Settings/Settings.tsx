/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BreadcrumbPage } from '@app/BreadcrumbPage/BreadcrumbPage';
import { FeatureFlag } from '@app/Shared/Components/FeatureFlag';
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { cleanDataId, getActiveTab, hashCode, switchTab } from '@app/utils/utils';
import {
  Card,
  Form,
  FormGroup,
  FormHelperText,
  HelperText,
  HelperTextItem,
  Label,
  Sidebar,
  SidebarContent,
  SidebarPanel,
  Tab,
  TabProps,
  Tabs,
  TabTitleText,
  Title,
} from '@patternfly/react-core';
import * as React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import { AutomatedAnalysis } from './Config/AutomatedAnalysis';
import { AutoRefresh } from './Config/AutoRefresh';
import { ChartCards } from './Config/ChartCards';
import { DatetimeControl } from './Config/DatetimeControl';
import { DeletionDialogControl } from './Config/DeletionDialogControl';
import { FeatureLevels } from './Config/FeatureLevels';
import { Language } from './Config/Language';
import { NotificationControl } from './Config/NotificationControl';
import { Theme } from './Config/Theme';
import { WebSocketDebounce } from './Config/WebSocketDebounce';
import { SettingGroup, SettingTab, _TransformedUserSetting } from './types';
import { paramAsTab, tabAsParam, getGroupFeatureLevel } from './utils';

export interface SettingsProps {}

export const Settings: React.FC<SettingsProps> = (_) => {
  const { t } = useTranslation();

  const settings = React.useMemo(
    () =>
      [
        NotificationControl,
        AutomatedAnalysis,
        ChartCards,
        DeletionDialogControl,
        WebSocketDebounce,
        AutoRefresh,
        FeatureLevels,
        Language,
        DatetimeControl,
        Theme,
      ].map(
        (c) =>
          ({
            title: t(c.titleKey) || '',
            description:
              typeof c.descConstruct === 'string' ? (
                t(c.descConstruct)
              ) : (
                // Use children prop to avoid i18n parses body as key
                /* eslint react/no-children-prop: 0 */
                <Trans i18nKey={c.descConstruct.key} children={c.descConstruct.parts} />
              ),
            element: React.createElement(c.content, null),
            category: c.category,
            disabled: c.disabled,
            orderInGroup: c.orderInGroup || -1,
            featureLevel: c.featureLevel || FeatureLevel.PRODUCTION,
          }) as _TransformedUserSetting,
      ),
    [t],
  );

  const navigate = useNavigate();
  const { search, pathname } = useLocation();

  const activeTab = React.useMemo(() => {
    return paramAsTab(
      getActiveTab(
        search,
        'tab',
        Object.values(SettingTab).map((v) => tabAsParam(v)),
        tabAsParam(SettingTab.GENERAL),
      ),
    );
  }, [search]);

  const onTabSelect = React.useCallback(
    (_: React.MouseEvent, key: string | number) =>
      switchTab(navigate, pathname, search, { tabKey: 'tab', tabValue: `${tabAsParam(key as SettingTab)}` }),
    [navigate, pathname, search],
  );

  const settingGroups = React.useMemo(() => {
    return Object.values(SettingTab).map((cat) => {
      const panels = settings.filter((s) => s.category === cat).sort((a, b) => b.orderInGroup - a.orderInGroup);
      return {
        groupLabel: t(cat),
        groupKey: cat,
        settings: panels,
        featureLevel: getGroupFeatureLevel(panels),
      };
    }) as SettingGroup[];
  }, [settings, t]);

  return (
    <>
      <BreadcrumbPage pageTitle="Settings">
        <Card isFullHeight>
          <Sidebar tabIndex={0}>
            <SidebarPanel>
              <Tabs
                isVertical
                isExpanded
                aria-label="Select setting category"
                activeKey={activeTab}
                onSelect={onTabSelect}
              >
                {settingGroups.map((grp) => (
                  <WrappedTab
                    key={`${grp.groupLabel}-tab`}
                    eventKey={grp.groupKey}
                    title={<TabTitleText>{grp.groupLabel}</TabTitleText>}
                    featureLevelConfig={{
                      level: grp.featureLevel,
                    }}
                  />
                ))}
              </Tabs>
            </SidebarPanel>
            <SidebarContent>
              {settingGroups
                .filter((grp) => grp.groupKey === activeTab)
                .map((grp) => (
                  <Form key={`${grp.groupKey}-setting`} className="settings__content">
                    {grp.settings.map((s, index) => (
                      <FeatureFlag level={s.featureLevel} key={`${grp.groupLabel}-${s.title}-${index}-flag`}>
                        <FormGroup
                          label={
                            <Title headingLevel={'h2'} size={'lg'}>
                              {s.title}
                              {s.featureLevel !== FeatureLevel.PRODUCTION && (
                                <Label
                                  isCompact
                                  style={{
                                    marginLeft: '1ch',
                                    textTransform: 'capitalize',
                                  }}
                                  color={s.featureLevel === FeatureLevel.BETA ? 'green' : 'red'}
                                >
                                  {FeatureLevel[s.featureLevel].toLowerCase()}
                                </Label>
                              )}
                            </Title>
                          }
                          key={`${grp.groupLabel}-${s.title}-${index}`}
                        >
                          <FormHelperText>
                            <HelperText>
                              <HelperTextItem>{s.description}</HelperTextItem>
                            </HelperText>
                          </FormHelperText>
                          {s.element}
                        </FormGroup>
                      </FeatureFlag>
                    ))}
                  </Form>
                ))}
            </SidebarContent>
          </Sidebar>
        </Card>
        <>
          {
            // Need this fragment to correct bottom margin.
          }
        </>
      </BreadcrumbPage>
    </>
  );
};

interface SettingTabProps extends TabProps {
  featureLevelConfig: {
    level: FeatureLevel;
    strict?: boolean;
  };
}

// Workaround to the Tabs component requiring children to be React.FC<TabProps>
const WrappedTab: React.FC<SettingTabProps> = ({ featureLevelConfig, eventKey, title, children }) => {
  const { t } = useTranslation();

  return (
    <FeatureFlag level={featureLevelConfig.level} strict={featureLevelConfig.strict}>
      <Tab
        eventKey={eventKey}
        title={title}
        id={`${hashCode(`${eventKey}`)}`}
        data-quickstart-id={cleanDataId(`settings-${t(eventKey.toString())}-tab`)}
      >
        {children}
      </Tab>
    </FeatureFlag>
  );
};

export default Settings;
