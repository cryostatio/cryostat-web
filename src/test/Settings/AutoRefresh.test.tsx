/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { AutoRefresh } from '@app/Settings/AutoRefresh';
import { defaultServices, ServiceContext } from '@app/Shared/Services/Services';
import { cleanup, screen } from '@testing-library/react';
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { renderWithServiceContext, testTranslate } from '../Common';

jest.spyOn(defaultServices.settings, 'autoRefreshEnabled').mockReturnValue(false);
jest.spyOn(defaultServices.settings, 'autoRefreshPeriod').mockReturnValue(30);
jest.spyOn(defaultServices.settings, 'autoRefreshUnits').mockReturnValue(1000);

describe('<AutoRefresh/>', () => {
  beforeEach(() => {
    // Clear mock instances, calls, and results
    jest.mocked(defaultServices.settings.setAutoRefreshEnabled).mockClear();
    jest.mocked(defaultServices.settings.setAutoRefreshPeriod).mockClear();
    jest.mocked(defaultServices.settings.setAutoRefreshUnits).mockClear();
  });
  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          {React.createElement(AutoRefresh.content, null)}
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should default to have auto-refresh disabled', async () => {
    renderWithServiceContext(React.createElement(AutoRefresh.content, null));

    const enableCheckbox = screen.getByLabelText('Enabled');
    expect(enableCheckbox).toBeInTheDocument();
    expect(enableCheckbox).toBeVisible();
    expect(enableCheckbox).not.toBeChecked();
  });

  it('should enable selections if checkbox is checked', async () => {
    const { user } = renderWithServiceContext(React.createElement(AutoRefresh.content, null));

    const enableCheckbox = screen.getByLabelText(testTranslate('SETTINGS.AUTO_REFRESH.CHECKBOX_LABEL'));
    expect(enableCheckbox).toBeInTheDocument();
    expect(enableCheckbox).toBeVisible();

    await user.click(enableCheckbox);

    [
      screen.getByLabelText('Duration Picker Period Input'),
      screen.getByLabelText('Duration Picker Units Input'),
    ].forEach((input) => {
      expect(input).toBeInTheDocument();
      expect(input).toBeVisible();
      expect(input).not.toBeDisabled();
    });

    expect(defaultServices.settings.setAutoRefreshEnabled).toHaveBeenCalledTimes(1);
    expect(defaultServices.settings.setAutoRefreshEnabled).toHaveBeenCalledWith(true);
  });

  it('should set value to local storage when congfigured', async () => {
    const { user } = renderWithServiceContext(React.createElement(AutoRefresh.content, null));

    const periodInput = screen.getByLabelText('Duration Picker Period Input');
    expect(periodInput).toBeInTheDocument();
    expect(periodInput).toBeVisible();

    const enableCheckbox = screen.getByLabelText(testTranslate('SETTINGS.AUTO_REFRESH.CHECKBOX_LABEL'));
    expect(enableCheckbox).toBeInTheDocument();
    expect(enableCheckbox).toBeVisible();

    await user.click(enableCheckbox);
    await user.clear(periodInput as HTMLInputElement);
    await user.type(periodInput, '20');

    // Clear + typing 2 + typing 0
    expect(defaultServices.settings.setAutoRefreshPeriod).toHaveBeenCalledTimes(3);
    expect(defaultServices.settings.setAutoRefreshPeriod).toHaveBeenLastCalledWith(20);

    const unitSelect = screen.getByLabelText('Duration Picker Units Input');
    expect(unitSelect).toBeInTheDocument();
    expect(unitSelect).toBeVisible();

    await user.selectOptions(unitSelect, 'Minutes');

    expect(defaultServices.settings.setAutoRefreshUnits).toHaveBeenCalledTimes(1);
    expect(defaultServices.settings.setAutoRefreshUnits).toHaveBeenLastCalledWith(60 * 1000);
  });
});
