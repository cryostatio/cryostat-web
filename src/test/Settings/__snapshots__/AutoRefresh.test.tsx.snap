// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoRefresh/> renders correctly 1`] = `
Array [
  <div
    className="pf-v5-l-split pf-m-gutter"
  >
    <div
      className="pf-v5-l-split__item pf-m-fill"
    >
      <span
        className="pf-v5-c-form-control pf-m-disabled"
      >
        <input
          aria-invalid={false}
          aria-label="Duration Picker Period Input"
          data-ouia-component-id="OUIA-Generated-TextInputBase-1"
          data-ouia-component-type="PF5/TextInput"
          data-ouia-safe={true}
          disabled={true}
          id="duration-picker-period"
          min="0"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="number"
          value={30}
        />
      </span>
    </div>
    <div
      className="pf-v5-l-split__item"
    >
      <span
        className="pf-v5-c-form-control pf-m-disabled"
      >
        <select
          aria-invalid={false}
          aria-label="Duration Picker Units Input"
          data-ouia-component-id="OUIA-Generated-FormSelect-default-1"
          data-ouia-component-type="PF5/FormSelect"
          data-ouia-safe={true}
          disabled={true}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          value={1000}
        >
          <option
            className=""
            disabled={false}
            value={1000}
          >
            Seconds
          </option>
          <option
            className=""
            disabled={false}
            value={60000}
          >
            Minutes
          </option>
          <option
            className=""
            disabled={false}
            value={3600000}
          >
            Hours
          </option>
        </select>
        <span
          className="pf-v5-c-form-control__utilities"
        >
          <span
            className="pf-v5-c-form-control__toggle-icon"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              />
            </svg>
          </span>
        </span>
      </span>
    </div>
  </div>,
  <div
    className="pf-v5-c-check"
  >
    <input
      aria-invalid={false}
      checked={false}
      className="pf-v5-c-check__input"
      data-ouia-component-id="OUIA-Generated-Checkbox-1"
      data-ouia-component-type="PF5/Checkbox"
      data-ouia-safe={true}
      data-quickstart-id="settings-connectivity-tab-auto-refresh"
      disabled={false}
      id="auto-refresh-enabled"
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
    <label
      className="pf-v5-c-check__label"
      htmlFor="auto-refresh-enabled"
    >
      Enabled
    </label>
  </div>,
]
`;
